ludwig/benchmarking/reporting.py

```diff
@@ -1,35 +1,166 @@
+    else:
+        raise ValueError(f"Device {kineto_event.device_type()} is not valid.")
+
+
+def get_device_memory_usage(
+    kineto_event: _KinetoEvent, mem_records_acc: profiler_util.MemRecordsAcc, run_usage_info: Dict[str, Any]
+) -> Dict[str, Any]:
+    """Get CPU and CUDA memory usage for an event.
+
+    :param kineto_event: a Kineto event instance.
+    :param mem_records_acc: memory events wrapped in a data structure for fast lookup.
+    :param run_usage_info: usage info for one execution of the tagged code block.
+    """
+    records_in_interval = mem_records_acc.in_interval(
+        kineto_event.start_us(), kineto_event.start_us() + kineto_event.duration_us()
+    )
+    memory_so_far = defaultdict(int)
+    count_so_far = defaultdict(int)
+    average_so_far = defaultdict(float)
+    max_so_far = defaultdict(int)
+
+    for mem_record in records_in_interval:
+        device, nbytes = get_memory_details(mem_record[0])
+        memory_so_far[device] += nbytes
+        max_so_far[device] = max(max_so_far[device], memory_so_far[device])
+        average_so_far[device] = (memory_so_far[device] + (average_so_far[device] * count_so_far[device])) / (
+            count_so_far[device] + 1
+        )
+        count_so_far[device] += 1
+    for device in count_so_far:
+        run_usage_info[f"torch_average_{device}_memory_usage"] = average_so_far[device]
+        run_usage_info[f"torch_max_{device}_memory_usage"] = max_so_far[device]
+    return run_usage_info
+
+
+def get_torch_op_time(events: List[profiler_util.FunctionEvent], attr: str) -> Union[int, float]:
+    """Get time torch operators spent executing for a list of events.
+
+    :param events: list of events.
+    :param attr: a FunctionEvent attribute. Expecting one of "cpu_time_total", "cuda_time_total".
+    """
+    if attr not in ["cpu_time_total", "cuda_time_total"]:
+        return -1
+
+    total = 0
+    for e in events:
+        if LUDWIG_TAG not in e.trace_name:
+            total += getattr(e, attr)
         else:
-            raise ValueError("Tag unrecognized. Please choose 'train' or 'evaluate'.")
+            total += get_torch_op_time(e.cpu_children, attr)
+    return total
+
+
+def get_device_run_durations(
+    function_event: profiler_util.FunctionEvent, run_usage_info: Dict[str, Any]
+) -> Dict[str, Any]:
+    """Get CPU and CUDA run durations for an event.
+
+    :param function_event: a function event instance.
+    :param run_usage_info: usage info for one execution of the tagged code block.
+    """
+    run_usage_info["torch_cpu_time"] = get_torch_op_time(function_event.cpu_children, "cpu_time_total")
+    run_usage_info["torch_cuda_time"] = get_torch_op_time(function_event.cpu_children, "cuda_time_total")
+    return run_usage_info
+
+
+def get_resource_usage_report(
+    main_kineto_events: List[_KinetoEvent],
+    main_function_events: List[profiler_util.FunctionEvent],
+    memory_events: List[Any],
+    info: Dict[str, Any],
+) -> Dict[str, Any]:
```

Comment:
If there's a finite number of keys, it would be worth defining a dataclass instead of using `Dict[str, Any]`.

Reply:
The number of keys and their names depend on how many times someone uses `LudwigProfiler`. So the size and keys of this dictionary can vary from run to another.